### PR TITLE
cli: new flag for JSON output of `constellation verify`

### PR DESCRIPTION
**Pull request by elchead**
### Context

Motivated by https://github.com/edgelesssys/constellation/pull/2357, this PR provides a JSON output format as communication interface between the CLI and the attestationconfigapi upload tool through a new flag `constellation verify --output json`.

### Proposed change(s)

- new internal pkg `verify` to provide the types for JSON serialization. 
- new `--output` flag to support 'json' for Azure
- moved `--raw` to `--output raw`

### Related issue
- [AB#3462](https://dev.azure.com/Edgeless/ae37573d-ccde-4af2-ab1e-001e587197d1/_workitems/edit/3462)

### Additional info
~To only obtain the JSON output needed for the upload tool, some lines need to be omitted:
`./constellation verify --force --json | tail -n +3 | head -n -1`~

### Checklist

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

**Comment by 3u13r** [comment]
This is not the implementation I suggested.
> implement a custom mashaller and unmarshaller which recursively parses the attestationDoc to a struct containing the parsed InstanceInfo

I said that we should implement a custom mashaller and unmarsaller. As we already do in e.g. https://github.com/edgelesssys/constellation/blob/cli/verify/json-output/internal/attestation/idkeydigest/idkeydigest.go#L180

Did you encounter any issue with this approach, or why did you prefer your implementation?

**Comment by daniel-weisse** [comment]
> To only obtain the JSON output needed for the upload tool, some lines need to be omitted

This seems like bad UX. If a command offers me the option to receive JSON output, I would expected all output to be JSON, not just some of it.
What about omitting the hints about retrieval of information from the id/state file, and potentially adding the verification status to the JSON output?
E.g.
```bash
$ constellation verify
{
  "attestationDocument": {
    // ...
  },
  "verification": "ok",
}
```

**Comment by thomasten** [comment]
> If a command offers me the option to receive JSON output, I would expected all output to be JSON, not just some of it.

A common way tools solve this is printing the JSON output to stdout and everything else to stderr.

**Comment by elchead** [comment]
> This is not the implementation I suggested.
> 
> > implement a custom mashaller and unmarshaller which recursively parses the attestationDoc to a struct containing the parsed InstanceInfo
> 
> I said that we should implement a custom mashaller and unmarsaller. As we already do in e.g. https://github.com/edgelesssys/constellation/blob/cli/verify/json-output/internal/attestation/idkeydigest/idkeydigest.go#L180
> 
> Did you encounter any issue with this approach, or why did you prefer your implementation?

I think if we wan

> > To only obtain the JSON output needed for the upload tool, some lines need to be omitted
> 
> This seems like bad UX. If a command offers me the option to receive JSON output, I would expected all output to be JSON, not just some of it. What about omitting the hints about retrieval of information from the id/state file, and potentially adding the verification status to the JSON output? E.g.
> 
> ```shell
> $ constellation verify
> {
>   "attestationDocument": {
>     // ...
>   },
>   "verification": "ok",
> }
> ```

I print to stderr now. Do we need the verification ok in the json output too? Since we don't foresee consuming this json in any other way besides the upload tool at the moment, I would drop it. An error of `verify` should indicate that the verification was not ok right?

**Comment by elchead** [comment]
> This is not the implementation I suggested.
> 
> > implement a custom mashaller and unmarshaller which recursively parses the attestationDoc to a struct containing the parsed InstanceInfo
> 
> I said that we should implement a custom mashaller and unmarsaller. As we already do in e.g. https://github.com/edgelesssys/constellation/blob/cli/verify/json-output/internal/attestation/idkeydigest/idkeydigest.go#L180
> 
> Did you encounter any issue with this approach, or why did you prefer your implementation?

I think it's better to use a proper type, because to unmarshal and read the struct we would need this anyways (as required by the upload tool)

**Comment by derpsteb** [comment]
Recap of our conversation on the phone:
- Ideally we could use a type from go-sev-guest that recursively parses all fields in an SNP report. However, looking at how scattered the bits of parsing logic currently are, it seems like considerably more effort to change the type upstream. It is also questionable if the maintainers would even want such a change.
- I personally don't think it is urgent to upstream this. It would be easy to adapt this PR to an upstream implementation, if one would be released/implemented tomorrow. 
- Using `newSNPReport` in the existing formatter would make sense to reduce code duplication.
- A point in the future when we _should_ try upstreaming this imo: when we plan on using the Report type in more places than attestationconfigapi and verify. 
- I opened an issue to see how the upstream maintainers feel about such a contribution. The answer can help inform our decision here: https://github.com/google/go-sev-guest/issues/91

File: cli/internal/cmd/verify.go
Comment (elchead) [reply cli/internal/cmd/verify.go:302]: > Is this a question for the reviewer or for yourself? Is this something we can decide and "fix" now?

I believe that we don't need the quote output in the json format and just wanted to confirm.

File: cli/internal/cmd/verify.go
Comment (elchead) [reply cli/internal/cmd/verify.go:89]: Most of the logic is now shared already through the parse SNP report. I remove this TODO.

**Review by 3u13r — DISMISSED** [review]
LGMT, I'd suggest to get a second approval for this though.

**Comment by elchead** [comment]
good catch, thanks! Since we also only support this for Azure on the default printer, I return an unsupported message:
```
./constellation verify -o json
Using endpoint from "constellation-id.json". Specify --node-endpoint to override this.
Using ID from "constellation-id.json". Specify --cluster-id to override this.
Error: printing attestation document: JSON output is currently only supported for Azure
```

File: cli/internal/cmd/verify.go
Comment (derpsteb) [inline cli/internal/cmd/verify.go:199]:
```suggestion
	c.log.Debugf("Flag 'output' set to %t", output)
```

and move to after the error check.

**Review by derpsteb — APPROVED** [review]
LGTM :+1:

**Review by 3u13r — CHANGES_REQUESTED** [review]

**Review by 3u13r — APPROVED** [review]
LGTM

**Pull request merged**
